Crash on Windows build

### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

I am using the latest version of master branch and I receive a crash due to exception throw:
```
SDL_main:  Exception 'No mapping for the Unicode character exists in the target multi-byte code page.' occurred during application runtime.
```

Run the game --> Load Game --> Standard

I believe this is related to recent #9098.

### Save file

None.

### Additional info

_No response_

Hi @oleg-derevenetz , #9117 fixes the crash. This is what I see in logs (which is fine):
```
22.09.2024 15:34:19: [ERROR]    StreamFile::open:  Error opening file C:\Users\sysadmin\AppData\Roaming\fheroes2\files\save\Great_War_II_0001_????.sav
22.09.2024 15:34:19: [DBG_GAME] Game::LoadSAV2FileInfo:  Error opening the file C:\Users\sysadmin\AppData\Roaming\fheroes2\files\save\Great_War_II_0001_????.sav
```

--- a/src/engine/audio.cpp
+++ b/src/engine/audio.cpp
@@ -246,7 +246,7 @@ namespace
             if ( std::holds_alternative<std::string>( _source ) ) {
                 const std::string & file = std::get<std::string>( _source );
 
-                std::unique_ptr<Mix_Music, void ( * )( Mix_Music * )> result( Mix_LoadMUS( System::encLocalToSDL( file ).c_str() ), Mix_FreeMusic );
+                std::unique_ptr<Mix_Music, void ( * )( Mix_Music * )> result( Mix_LoadMUS( System::encLocalToUTF8( file ).c_str() ), Mix_FreeMusic );
                 if ( !result ) {
                     ERROR_LOG( "Failed to create a music track from file " << file << ". The error: " << Mix_GetError() )
                 }
@@ -1096,7 +1096,7 @@ void Music::SetMidiSoundFonts( const ListFiles & files )
         filePaths.pop_back();
     }
 
-    if ( Mix_SetSoundFonts( System::encLocalToSDL( filePaths ).c_str() ) == 0 ) {
+    if ( Mix_SetSoundFonts( System::encLocalToUTF8( filePaths ).c_str() ) == 0 ) {
         ERROR_LOG( "Failed to set MIDI SoundFonts using paths " << filePaths << ". The error: " << Mix_GetError() )
     }
 }

--- a/src/engine/dir.cpp
+++ b/src/engine/dir.cpp
@@ -85,7 +85,7 @@ namespace
                 continue;
             }
 
-            files.emplace_back( System::fsPathToString( entryPath.string() ) );
+            files.emplace_back( System::fsPathToString( entryPath ) );
         }
     }
 }

--- a/src/engine/dir.cpp
+++ b/src/engine/dir.cpp
@@ -81,21 +81,11 @@ namespace
 
             const std::filesystem::path & entryPath = entry.path();
 
-            if (
-#if defined( _WIN32 )
-                const std::string fileName = System::encUTF8ToLocal( entryPath.filename().u8string() );
-#else
-                const std::string fileName = entryPath.filename().string();
-#endif
-                !nameFilter( fileName, needExactMatch, filter, strCmp ) ) {
+            if ( !nameFilter( System::fsPathToString( entryPath.filename() ), needExactMatch, filter, strCmp ) ) {
                 continue;
             }
 
-#if defined( _WIN32 )
-            files.emplace_back( System::encUTF8ToLocal( entryPath.u8string() ) );
-#else
-            files.emplace_back( entryPath.string() );
-#endif
+            files.emplace_back( System::fsPathToString( entryPath.string() ) );
         }
     }
 }

--- a/src/engine/dir.cpp
+++ b/src/engine/dir.cpp
@@ -79,12 +79,23 @@ namespace
                 continue;
             }
 
-            const std::string fileName = entry.path().filename().string();
-            if ( !nameFilter( fileName, needExactMatch, filter, strCmp ) ) {
+            const std::filesystem::path & entryPath = entry.path();
+
+            if (
+#if defined( _WIN32 )
+                const std::string fileName = System::encUTF8ToLocal( entryPath.filename().u8string() );
+#else
+                const std::string fileName = entryPath.filename().string();
+#endif
+                !nameFilter( fileName, needExactMatch, filter, strCmp ) ) {
                 continue;
             }
 
-            files.emplace_back( System::concatPath( correctedPath, fileName ) );
+#if defined( _WIN32 )
+            files.emplace_back( System::encUTF8ToLocal( entryPath.u8string() ) );
+#else
+            files.emplace_back( entryPath.string() );
+#endif
         }
     }
 }

--- a/src/engine/image_tool.cpp
+++ b/src/engine/image_tool.cpp
@@ -125,17 +125,17 @@ namespace
         int res = 0;
 
         if ( isPNGFilePath( path ) ) {
-            res = IMG_SavePNG( surface.get(), System::encLocalToSDL( path ).c_str() );
+            res = IMG_SavePNG( surface.get(), System::encLocalToUTF8( path ).c_str() );
         }
         else {
-            res = SDL_SaveBMP( surface.get(), System::encLocalToSDL( path ).c_str() );
+            res = SDL_SaveBMP( surface.get(), System::encLocalToUTF8( path ).c_str() );
         }
 #else
         if ( isPNGFilePath( path ) ) {
             memcpy( path.data() + path.size() - 3, "bmp", 3 );
         }
 
-        const int res = SDL_SaveBMP( surface.get(), System::encLocalToSDL( path ).c_str() );
+        const int res = SDL_SaveBMP( surface.get(), System::encLocalToUTF8( path ).c_str() );
 #endif
 
         return res == 0;
@@ -177,9 +177,9 @@ namespace fheroes2
         std::unique_ptr<SDL_Surface, void ( * )( SDL_Surface * )> loadedSurface( nullptr, SDL_FreeSurface );
 
 #if defined( WITH_IMAGE )
-        loadedSurface.reset( IMG_Load( System::encLocalToSDL( path ).c_str() ) );
+        loadedSurface.reset( IMG_Load( System::encLocalToUTF8( path ).c_str() ) );
 #else
-        loadedSurface.reset( SDL_LoadBMP( System::encLocalToSDL( path ).c_str() ) );
+        loadedSurface.reset( SDL_LoadBMP( System::encLocalToUTF8( path ).c_str() ) );
 #endif
         if ( !loadedSurface ) {
             return false;

--- a/src/engine/screen.cpp
+++ b/src/engine/screen.cpp
@@ -912,7 +912,7 @@ namespace
                 return;
             }
 
-            SDL_SetWindowTitle( _window, System::encLocalToSDL( title ).c_str() );
+            SDL_SetWindowTitle( _window, System::encLocalToUTF8( title ).c_str() );
         }
 
         void setIcon( const fheroes2::Image & icon ) override
@@ -1000,7 +1000,7 @@ namespace
                 if ( !isFullScreen() ) {
                     SDL_GetWindowPosition( _window, &_prevWindowPos.x, &_prevWindowPos.y );
                 }
-                _previousWindowTitle = System::encSDLToLocal( SDL_GetWindowTitle( _window ) );
+                _previousWindowTitle = System::encUTF8ToLocal( SDL_GetWindowTitle( _window ) );
 
                 SDL_DestroyWindow( _window );
                 _window = nullptr;
@@ -1104,7 +1104,7 @@ namespace
 
             flags |= SDL_WINDOW_RESIZABLE;
 
-            _window = SDL_CreateWindow( System::encLocalToSDL( _previousWindowTitle ).c_str(), _prevWindowPos.x, _prevWindowPos.y, resolutionInfo.screenWidth,
+            _window = SDL_CreateWindow( System::encLocalToUTF8( _previousWindowTitle ).c_str(), _prevWindowPos.x, _prevWindowPos.y, resolutionInfo.screenWidth,
                                         resolutionInfo.screenHeight, flags );
             if ( _window == nullptr ) {
                 ERROR_LOG( "Failed to create an application window of " << resolutionInfo.screenWidth << " x " << resolutionInfo.screenHeight

--- a/src/engine/system.cpp
+++ b/src/engine/system.cpp
@@ -361,11 +361,11 @@ std::string System::GetConfigDirectory( const std::string_view appName )
     std::string result = [&appName]() -> std::string {
 #if defined( __linux__ ) && !defined( ANDROID )
         if ( const char * configEnv = getenv( "XDG_CONFIG_HOME" ); configEnv != nullptr ) {
-            return System::concatPath( configEnv, appName );
+            return concatPath( configEnv, appName );
         }
 
         if ( const char * homeEnv = getenv( "HOME" ); homeEnv != nullptr ) {
-            return System::concatPath( System::concatPath( homeEnv, ".config" ), appName );
+            return concatPath( concatPath( homeEnv, ".config" ), appName );
         }
 
         return { "." };
@@ -394,17 +394,17 @@ std::string System::GetDataDirectory( const std::string_view appName )
     std::string result = [&appName]() -> std::string {
 #if defined( __linux__ ) && !defined( ANDROID )
         if ( const char * dataEnv = getenv( "XDG_DATA_HOME" ); dataEnv != nullptr ) {
-            return System::concatPath( dataEnv, appName );
+            return concatPath( dataEnv, appName );
         }
 
         if ( const char * homeEnv = getenv( "HOME" ); homeEnv != nullptr ) {
-            return System::concatPath( System::concatPath( homeEnv, ".local/share" ), appName );
+            return concatPath( concatPath( homeEnv, ".local/share" ), appName );
         }
 
         return { "." };
 #elif defined( MACOS_APP_BUNDLE )
         if ( const char * homeEnv = getenv( "HOME" ); homeEnv != nullptr ) {
-            return System::concatPath( System::concatPath( homeEnv, "Library/Application Support" ), appName );
+            return concatPath( concatPath( homeEnv, "Library/Application Support" ), appName );
         }
 
         return { "." };

--- a/src/engine/system.cpp
+++ b/src/engine/system.cpp
@@ -25,7 +25,6 @@
 
 #include <cassert>
 #include <cstdlib>
-#include <filesystem>
 #include <functional>
 #include <initializer_list>
 #include <map>
@@ -613,21 +612,11 @@ void System::globFiles( const std::string_view glob, std::vector<std::string> & 
     for ( const std::filesystem::directory_entry & entry : std::filesystem::directory_iterator( dirPath, ec ) ) {
         const std::filesystem::path & entryPath = entry.path();
 
-        if (
-#if defined( _WIN32 )
-            const std::string fileName = encUTF8ToLocal( entryPath.filename().u8string() );
-#else
-            const std::string fileName = entryPath.filename().string();
-#endif
-            !globMatch( fileName, pattern ) ) {
+        if ( !globMatch( fsPathToString( entryPath.filename() ), pattern ) ) {
             continue;
         }
 
-#if defined( _WIN32 )
-        fileNames.emplace_back( encUTF8ToLocal( entryPath.u8string() ) );
-#else
-        fileNames.emplace_back( entryPath.string() );
-#endif
+        fileNames.emplace_back( fsPathToString( entryPath.string() ) );
 
         isNoMatches = false;
     }
@@ -652,6 +641,19 @@ std::string System::encUTF8ToLocal( const std::string_view str )
     return convertBetweenACPAndUTF8( str, EncodingConversionDirection::UTF8ToACP );
 #else
     return std::string{ str };
+#endif
+}
+
+std::string System::fsPathToString( const std::filesystem::path & path )
+{
+#if defined( _WIN32 )
+    // On Windows, std::filesystem::path::string() can throw an exception if path contains UTF-16 characters that
+    // are non-representable in CP_ACP. However, converting a well-formed UTF-16 string to UTF-8 is always safe,
+    // so we perform this conversion first, and then convert the resulting UTF-8 to CP_ACP using our conversion
+    // function, which can never throw an exception.
+    return encUTF8ToLocal( path.u8string() );
+#else
+    return path.string();
 #endif
 }
 

--- a/src/engine/system.cpp
+++ b/src/engine/system.cpp
@@ -616,7 +616,7 @@ void System::globFiles( const std::string_view glob, std::vector<std::string> & 
             continue;
         }
 
-        fileNames.emplace_back( fsPathToString( entryPath.string() ) );
+        fileNames.emplace_back( fsPathToString( entryPath ) );
 
         isNoMatches = false;
     }

--- a/src/engine/system.cpp
+++ b/src/engine/system.cpp
@@ -120,8 +120,8 @@ namespace
         }
 
 #if SDL_VERSION_ATLEAST( 2, 0, 1 )
-        if ( const std::unique_ptr<char, void ( * )( void * )> path( SDL_GetPrefPath( "", System::encLocalToSDL( std::string{ appName } ).c_str() ), SDL_free ); path ) {
-            return System::encSDLToLocal( path.get() );
+        if ( const std::unique_ptr<char, void ( * )( void * )> path( SDL_GetPrefPath( "", System::encLocalToUTF8( std::string{ appName } ).c_str() ), SDL_free ); path ) {
+            return System::encUTF8ToLocal( path.get() );
         }
 #endif
 
@@ -613,11 +613,23 @@ void System::globFiles( const std::string_view glob, std::vector<std::string> & 
     for ( const std::filesystem::directory_entry & entry : std::filesystem::directory_iterator( dirPath, ec ) ) {
         const std::filesystem::path & entryPath = entry.path();
 
-        if ( globMatch( entryPath.filename().string(), pattern ) ) {
-            fileNames.push_back( entryPath.string() );
-
-            isNoMatches = false;
+        if (
+#if defined( _WIN32 )
+            const std::string fileName = encUTF8ToLocal( entryPath.filename().u8string() );
+#else
+            const std::string fileName = entryPath.filename().string();
+#endif
+            !globMatch( fileName, pattern ) ) {
+            continue;
         }
+
+#if defined( _WIN32 )
+        fileNames.emplace_back( encUTF8ToLocal( entryPath.u8string() ) );
+#else
+        fileNames.emplace_back( entryPath.string() );
+#endif
+
+        isNoMatches = false;
     }
 
     if ( isNoMatches ) {
@@ -625,7 +637,7 @@ void System::globFiles( const std::string_view glob, std::vector<std::string> & 
     }
 }
 
-std::string System::encLocalToSDL( const std::string_view str )
+std::string System::encLocalToUTF8( const std::string_view str )
 {
 #if defined( _WIN32 )
     return convertBetweenACPAndUTF8( str, EncodingConversionDirection::ACPToUTF8 );
@@ -634,7 +646,7 @@ std::string System::encLocalToSDL( const std::string_view str )
 #endif
 }
 
-std::string System::encSDLToLocal( const std::string_view str )
+std::string System::encUTF8ToLocal( const std::string_view str )
 {
 #if defined( _WIN32 )
     return convertBetweenACPAndUTF8( str, EncodingConversionDirection::UTF8ToACP );

--- a/src/engine/system.h
+++ b/src/engine/system.h
@@ -25,6 +25,7 @@
 #define H2SYSTEM_H
 
 #include <ctime>
+#include <filesystem>
 #include <string>
 #include <string_view>
 #include <vector>
@@ -72,6 +73,9 @@ namespace System
     // Windows this app works with ANSI code page (CP_ACP). On all other systems, it is currently assumed that UTF-8 encoding
     // is used, and the original string is simply returned unchanged.
     std::string encUTF8ToLocal( const std::string_view str );
+
+    // Performs a safe conversion of an std::filesystem::path instance to a string. See the implementation for details.
+    std::string fsPathToString( const std::filesystem::path & path );
 
     tm GetTM( const time_t time );
 }

--- a/src/engine/system.h
+++ b/src/engine/system.h
@@ -63,15 +63,15 @@ namespace System
     // are no files matching the pattern, it will be appended to the 'fileNames' as is.
     void globFiles( const std::string_view glob, std::vector<std::string> & fileNames );
 
-    // Converts the given string from local encoding to SDL encoding (UTF-8). It is used mainly on Windows, because on Windows
-    // this app works with ANSI code page (CP_ACP). On all other systems, it is currently assumed that UTF-8 encoding is used,
-    // and the original string is simply returned unchanged.
-    std::string encLocalToSDL( const std::string_view str );
+    // Converts the given string from local encoding to UTF-8 encoding (used by SDL). It is used mainly on Windows, because on
+    // Windows this app works with ANSI code page (CP_ACP). On all other systems, it is currently assumed that UTF-8 encoding
+    // is used, and the original string is simply returned unchanged.
+    std::string encLocalToUTF8( const std::string_view str );
 
-    // Converts the given string from SDL encoding (UTF-8) to local encoding. It is used mainly on Windows, because on Windows
-    // this app works with ANSI code page (CP_ACP). On all other systems, it is currently assumed that UTF-8 encoding is used,
-    // and the original string is simply returned unchanged.
-    std::string encSDLToLocal( const std::string_view str );
+    // Converts the given string from UTF-8 encoding (used by SDL) to local encoding. It is used mainly on Windows, because on
+    // Windows this app works with ANSI code page (CP_ACP). On all other systems, it is currently assumed that UTF-8 encoding
+    // is used, and the original string is simply returned unchanged.
+    std::string encUTF8ToLocal( const std::string_view str );
 
     tm GetTM( const time_t time );
 }

